docs(CSN-568): Update main readme (#288)
neuralinternet/compute-subnet: README.md

---

## Miner vs. Validator Overview
![Miner Overview Diagram](docs/sn27_miner_overview1.png)

### Miner
- **Role**: Contribute resources (GPU instances) to the network.  
- **Rewards**: Performance-based. Higher-performance devices with more GPUs receive higher rewards.  
- **Key Requirements**:  
  - GPU(s) with up-to-date drivers.  
  - Properly opened ports. E.g. 4444 for validator allocations, 8091 for serving the axon.  
  - A registered wallet hotkey on the correct Bittensor subnet (netuid 27 for main, or netuid 15 for test).

### Validator
- **Role**: Verifies the miners' computational integrity, assigns performance scores, and dynamically allocates resources to clients.  
- **Actions**:  
  - Requests performance data (e.g., GPU type, memory) from miners.  
  - Benchmarks or runs tasks to confirm advertised hardware.  
  - Updates scores that determine miners’ reward weights.  
- **Key Requirements**:  
  - A registered wallet hotkey on the correct Bittensor subnet (netuid 27 for main, or netuid 15 for test).
  - Up-to-date code to ensure accurate scoring.

---

## Installation and Prerequisites

### System Requirements
- **Operating System**: Ubuntu 22.04 (recommended).

> - **Python**: 3.8 or higher.

Reviewer: I think we should make py3.10 the minimal version, we are not making a general-purpose dev library here it doesn't make sense to support such old versions. 3.10 included some important features everybody use nowadays (I think I saw a few snippets that would actually misbehave if actually tried on py3.8) and Ubuntu 22.04 definitely included it anyway.

Author: on it
thanks

- **GPU**: NVIDIA GPU (recommended).

> **Important**: Each UID is limited to **one external IP**. **Port 4444 is to be opened** for your miner to be allocated properly. Automatic blacklisting occurs for anomalous behavior.

### Install Docker
A Docker environment is required for miner resource allocation:

1. **Install Docker** on Ubuntu:  
   [Official Docs](https://docs.docker.com/engine/install/ubuntu)

2. **Verify Docker**:
   ```bash
   docker run hello-world
   ```
   This should display a confirmation message with no errors.

### Install Bittensor
1. **Run the one-line installer**:
   ```bash
   /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/opentensor/bittensor/master/scripts/install.sh)"
   ```
2. **Verify Installation**:
   ```bash
   btcli --help
   ```
   If you get a command-not-found error, add Bittensor to your PATH:
   ```bash

>    export PATH=$PATH:$(python3 -m site --user-base)/bin

Reviewer: This is ephemeral and a line should be added to bashrc or someplace like that to persist (like in other examples).

|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
```
```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Fri_Nov__3_17:16:49_PDT_2023
Cuda compilation tools, release 12.3, V12.3.103
Build cuda_12.3.r12.3/compiler.33492891_0
```

### NVIDIA Docker Support
Enable GPU functionality within Docker containers:

```bash
distribution=$(. /etc/os-release;echo $ID$VERSION_ID)
curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list \
  | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
sudo apt update
sudo apt-get install -y nvidia-container-toolkit
sudo apt install -y nvidia-docker2
```

### WandB Setup
1. **Create a free WandB account**: [wandb.ai](https://wandb.ai/)  
2. **Obtain an API Key** and place it into your `.env` file:

>    ```bash
>    cd Compute-Subnet
>    mv .env.example .env
>    # Use VIM or another text aditor to replace the placeholder with your actual API key
>    ```

Reviewer: I'm thinking let's make it less error-prone (and also I think directory name is incorrectly cased), there are more safe ways to copy-paste a key and if we ever need to extend the list we'd just add more prompts. I think this:
```suggestion
   ```bash
   cd compute-subnet
   read -p "Enter WanDB API key: " wandb_api_key && sed -e s/"your_api_key"/"${wandb_api_key}"/ .env.example > .env
   ```
```

    legacy (l)          Miscellaneous commands.
    info (i)            Instructions for enabling autocompletion for the CLI.

options:
  -h, --help            show this help message and exit
  --print-completion {bash,zsh,tcsh}
                        Print shell tab completion script
```                       
See Bittensor’s documentation for alternative installation instructions.
[Bittensor Documentation](https://docs.bittensor.com/)  

### Create or Regenerate Keys
1. **Create new coldkey** (stores funds):
   ```bash
   btcli w new_coldkey
   ```
2. **Create new hotkey** (used for daily operations e.g. mining/validating/registration):
   ```bash
   btcli w new_hotkey
   ```
3. **Regenerate existing keys** if needed (to import them on this machine):
   ```bash
   btcli w regen_coldkeypub #see below
   btcli w regen_coldkey
   btcli w regen_hotkey
   ```

> **Tip**: For security, you can generate your coldkey on a secure offline machine and only load the public portion onto your miner or validator servers.

### Clone and Install Compute Subnet
1. **Clone the repository**:
   ```bash
   git clone https://github.com/neuralinternet/compute-subnet.git

>    cd Compute-Subnet

Reviewer: casing is important
```suggestion
   cd compute-subnet
```

   ```
2. **Install dependencies**:
   ```bash
   python3 -m pip install -r requirements.txt
   python3 -m pip install --no-deps -r requirements-compute.txt
   python3 -m pip install -e .
   ```
3. **In case you have missing requirements**:
```
sudo apt -y install ocl-icd-libopencl1 pocl-opencl-icd
```

### CUDA Toolkit and GPU Drivers
> **Tip**: If Nvidia toolkit and drivers are already installed on your machine, scroll down to step 5 to verify then move on to the docker CUDA support.
1. **Download** the latest CUDA from [NVIDIA's official page](https://developer.nvidia.com/cuda-downloads).  
2. **Install** (example for Ubuntu 22.04 (Dec. 2024)):
   ```bash
   wget https://developer.download.nvidia.com/compute/cuda/12.3.1/local_installers/cuda-repo-ubuntu2204-12-3-local_12.3.1-545.23.08-1_amd64.deb
   sudo dpkg -i cuda-repo-ubuntu2204-12-3-local_12.3.1-545.23.08-1_amd64.deb
   sudo cp /var/cuda-repo-ubuntu2204-12-3-local/cuda-*-keyring.gpg /usr/share/keyrings/
   sudo apt-get update
   sudo apt-get -y install cuda-toolkit-12-3
   sudo apt-get -y install cuda-drivers
   ```
3. **Set environment variables**:

>    ```bash
>    export CUDA_VERSION=cuda-12.3
>    export PATH=$PATH:/usr/local/$CUDA_VERSION/bin
>    export LD_LIBRARY_PATH=/usr/local/$CUDA_VERSION/lib64
>    echo "" >> ~/.bashrc
>    echo "PATH=$PATH" >> ~/.bashrc
>    echo "LD_LIBRARY_PATH=$LD_LIBRARY_PATH" >> ~/.bashrc
>    source ~/.bashrc
>    ```

Reviewer: this might be potentially unsafe and is a little redundant on the first run and wouldn't actually correctly export the variables from bashrc
```suggestion
   ```bash
   echo "" >> ~/.bashrc
   echo 'export CUDA_VERSION=cuda-12.3' >> ~/.bashrc
   echo 'export PATH="$PATH:/usr/local/$CUDA_VERSION/bin"' >> ~/.bashrc
   echo 'export LD_LIBRARY_PATH="/usr/local/$CUDA_VERSION/lib64:$LD_LIBRARY_PATH"' >> ~/.bashrc
   source ~/.bashrc
   ```
```

-- this should have the same initial effect in fewer lines plus proper persistence (previous version notably lacked export in bashrc) plus safer configuration (shouldn't break any other changes to PATH or LD_LIBRARY_PATH made earlier in bash init sequence)

   sudo systemctl start docker
   sudo apt install at  # ensures certain dependencies for Docker
   sudo service docker status  # should show 'active (running)'
   ```

---

## Registering Your Hotkey
You need **$TAO** tokens in your coldkey to register the hotkey on the chosen netuid.

- **Main Network** (netuid = 27, known as `finney`):
  ```bash
  btcli s register --subtensor.network finney --netuid 27
  ```
- **Test Network** (netuid = 15, known as `test`):
  ```bash
  btcli s register --subtensor.network test --netuid 15
  ```

> If you get the error `too many registrations this interval`, wait for the next interval and retry.  
> **Registration cost** can be checked [Here](https://taostats.io/subnets/netuid-27/#registration). Or check using the CLI w/ `btcli s list` (Cost is 'RECYCLE').

---

## Running a Miner
The miner contributes GPU resources to the network. Make sure your ports are open and your environment is set up as described above.

**General Miner Command**:
```bash
pm2 start ./neurons/miner.py --name <MINER_NAME> --interpreter python3 -- \
  --netuid 27 \
  --subtensor.network finney \
  --wallet.name <COLDKEY_NAME> \
  --wallet.hotkey <HOTKEY_NAME> \

>   --axon.port <xxxx> \

Reviewer: Let's simplify visually
```suggestion
  --axon.port 8091 \
```

  --logging.debug
```

- **`--netuid`**: Subnet ID (27 for main, 15 for test).
- **`--subtensor.network`**: Your Bittensor chain endpoint.  
  - **Main**: `finney`
  - **Test**: `test`  
  - Or use a custom endpoint, e.g. `subvortex.info:9944` (recommended)

> - **`--wallet.name`** & **`--wallet.hotkey`**: The coldkey/hotkey names you created.

Reviewer: ```suggestion
- **`--wallet.name`** & **`--wallet.hotkey`**: The coldkey/hotkey names you created [above](#create-or-regenerate-keys) and used in registration (btcli's defaults are `default` and `default` but both can be freely customized)
```
this should be a little better because neither is explicitly mentioned in the document before (they are prompted by btcli but would have to be explicit here)

> - **`--axon.port`**: A port opened with UFW as instructed [above](#networking-and-firewall) (e.g., 8091) to serve your axon. Important for proper functionality and miner<->validator communication.

Reviewer: ```suggestion
- **`--axon.port`**: default 8091 can be replaced with any port number allowed by ufw as instructed [above](#networking-and-firewall) to serve your axon. Important for proper functionality and miner<->validator communication.
```

   sudo apt install ufw
   sudo ufw allow 4444
   sudo ufw allow 22/tcp
   #below is an example axon port

>    #sudo ufw allow 8091/tcp

Reviewer: I think it's okay to uncomment this, copy-pasting without changes should result in valid default configuration (as long as it's clear what can be customized by advanced users)
